### PR TITLE
Reduce one allocation in spawning a future

### PR DESCRIPTION
**Pull request by sticnarf**
It needs two allocations to create a future `TaskCell` (one for the boxed future and one for the `Arc`). But in theory, only one allocation is needed, we can put the future together with the reference counter.

This PR is an attemp to reduce one allocation when spawning a future. 

## Benchmark

The benchmark result varies a bit on AMD and Intel CPUs. The improvement on Intel CPUs is larger. But generally the performance improves a lot.

Intel CPU (Broadwell architecture):

```
hained_spawn/yatp::future/256                                                                            
                        time:   [320.11 us 325.44 us 330.77 us]
                        change: [-16.766% -14.984% -13.236%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future::multilevel/256                                                                            
                        time:   [714.11 us 721.86 us 729.12 us]
                        change: [-9.1268% -7.6689% -6.2970%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future/512                                                                            
                        time:   [645.12 us 654.96 us 664.91 us]
                        change: [-13.688% -10.933% -7.4314%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future::multilevel/512                                                                             
                        time:   [1.4204 ms 1.4364 ms 1.4530 ms]
                        change: [-9.2218% -7.9877% -6.8580%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future/1024                                                                             
                        time:   [1.2643 ms 1.2847 ms 1.3045 ms]
                        change: [-17.206% -15.192% -13.257%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future::multilevel/1024                                                                             
                        time:   [2.7879 ms 2.8175 ms 2.8481 ms]
                        change: [-5.8462% -4.2189% -2.4824%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future/256                                                                             
                        time:   [541.62 us 562.19 us 584.14 us]
                        change: [-17.698% -13.384% -9.1931%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/256                                                                             
                        time:   [934.63 us 989.69 us 1.0481 ms]
                        change: [-14.833% -7.3358% +0.8149%] (p = 0.08 > 0.05)
                        No change in performance detected.

ping_pong/yatp::future/512                                                                             
                        time:   [1.1379 ms 1.1989 ms 1.2707 ms]
                        change: [-15.509% -10.600% -4.8128%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/512                                                                            
                        time:   [1.5878 ms 1.7102 ms 1.8519 ms]
                        change: [-11.634% -3.1553% +7.2989%] (p = 0.51 > 0.05)
                        No change in performance detected.

ping_pong/yatp::future/1024                                                                             
                        time:   [1.8029 ms 1.8493 ms 1.8967 ms]
                        change: [-25.694% -21.901% -17.945%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/1024                                                                            
                        time:   [3.2063 ms 3.4405 ms 3.7010 ms]
                        change: [-20.380% -11.576% -1.4083%] (p = 0.03 < 0.05)
                        Performance has improved.
```

AMD CPU (Zen3):

```
chained_spawn/yatp::future/256                                                                            
                        time:   [135.65 us 138.50 us 140.52 us]
                        change: [-5.9478% -2.5318% +1.0341%] (p = 0.17 > 0.05)
                        No change in performance detected.

chained_spawn/yatp::future::multilevel/256                                                                            
                        time:   [347.79 us 348.56 us 349.37 us]
                        change: [-5.1074% -3.4979% -2.0763%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future/512                                                                            
                        time:   [268.32 us 271.67 us 274.19 us]
                        change: [-9.1894% -7.1292% -5.0224%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future::multilevel/512                                                                            
                        time:   [682.44 us 684.55 us 686.96 us]
                        change: [-4.4316% -2.6463% -1.0859%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future/1024                                                                            
                        time:   [534.57 us 539.47 us 543.10 us]
                        change: [-9.5112% -7.3240% -5.4031%] (p = 0.00 < 0.05)
                        Performance has improved.

chained_spawn/yatp::future::multilevel/1024                                                                             
                        time:   [1.3547 ms 1.3590 ms 1.3634 ms]
                        change: [-4.1776% -2.9282% -1.8075%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future/256                                                                            
                        time:   [160.72 us 164.39 us 168.04 us]
                        change: [-20.615% -17.699% -14.543%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/256                                                                            
                        time:   [938.01 us 944.54 us 951.49 us]
                        change: [-8.5092% -6.3584% -4.6114%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future/512                                                                            
                        time:   [299.33 us 304.79 us 310.39 us]
                        change: [-19.910% -17.778% -15.550%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/512                                                                             
                        time:   [1.7924 ms 1.8224 ms 1.8491 ms]
                        change: [-12.183% -10.088% -8.4730%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future/1024                                                                            
                        time:   [580.94 us 591.65 us 601.87 us]
                        change: [-19.182% -17.125% -15.031%] (p = 0.00 < 0.05)
                        Performance has improved.

ping_pong/yatp::future::multilevel/1024                                                                             
                        time:   [3.8410 ms 3.8641 ms 3.8859 ms]
                        change: [-6.3105% -5.3388% -4.1879%] (p = 0.00 < 0.05)
                        Performance has improved.
```

### CODE REVIEW

File: src/task/future.rs [hunk 1]
```diff
@@ -24,19 +28,71 @@ struct TaskExtras {
...
+    fn extras(&self) -> &UnsafeCell<TaskExtras> {
+        unsafe { &self.0.as_ref().extras }
+    }
+
+    unsafe fn poll(&self, cx: &mut Context<'_>) -> Poll<()> {
+        (self.0.as_ref().poll_fn)(self, cx)
+    }
+
+    fn into_raw(self) -> *const () {
+        self.0.as_ptr() as _
```
Comment (BusyJay) [inline src/task/future.rs:107]: I think `mem::forget` is needed here.

**Comment by sticnarf** [comment]
@BusyJay Please take a look when free. I add the related benchmark result now.

**Review by BusyJay — APPROVED** [review]
Well done! Perhaps we need sanitizers to guarantee complex refactor will not bring subtle safety problems.

**Comment by sticnarf** [comment]
@BusyJay I add address sanitizer to CI in https://github.com/tikv/yatp/pull/57

**Pull request merged**